File: python/world.py
import pickle
import random

import pygame
from pygame.locals import *

from utils.constants import GRID_HEIGHT, GRID_WIDTH
from gui.display import Display
from grid import Grid
from gui.grid_map import GridMap
from gui.logger import Logger
from organisms.species import Species

from organisms.animals import Human, Wolf, Sheep, Fox, Turtle, Antelope, CyberSheep
from organisms.plants import Grass, SowThistle, Guarana, Belladonna, SosnowskysHogweed
from utils.point import Point


class World:
    VALID_KEYS = (K_a, K_w, K_s, K_d)

    @staticmethod
    def create_organism(species, position, grid):
        organism = None
        match species:
            case Species.HUMAN:
                organism = Human(position)
            case Species.WOLF:
                organism = Wolf(position)
            case Species.SHEEP:
                organism = Sheep(position)
            case Species.FOX:
                organism = Fox(position)
            case Species.TURTLE:
                organism = Turtle(position)
            case Species.ANTELOPE:
                organism = Antelope(position)
            case Species.CYBER_SHEEP:
                organism = CyberSheep(position)
            case Species.GRASS:
                organism = Grass(position)
            case Species.SOW_THISTLE:
                organism = SowThistle(position)
            case Species.GUARANA:
                organism = Guarana(position)
            case Species.BELLADONNA:
                organism = Belladonna(position)
            case Species.SOSNOWSKYS_HOGWEED:
                organism = SosnowskysHogweed(position)

        grid.add_organism(organism)

    def __init__(self):
        self.turn = 1
        self.grid = Grid(GRID_WIDTH, GRID_HEIGHT)
        self.grid_map = GridMap(self.grid)
        self.screen = Display(self.grid_map)
        self.unoccupied_points = self.grid.get_all_points()
        self.player = Human(self.pop_random_point())

    def init(self):
        self.grid.add_organism(self.player)

        self.add_organism(Species.SOSNOWSKYS_HOGWEED)
        self.add_organism(Species.SOSNOWSKYS_HOGWEED)
        self.add_organism(Species.SOSNOWSKYS_HOGWEED)
        self.add_organism(Species.SOSNOWSKYS_HOGWEED)
        self.add_organism(Species.SOSNOWSKYS_HOGWEED)
        self.add_organism(Species.CYBER_SHEEP)
        self.add_organism(Species.CYBER_SHEEP)
        # self.add_each_organism()

    def add_each_organism(self):
        [self.add_organism(species) for species in Species if species != Species.HUMAN]

    def add_organism(self, species):
        self.create_organism(species, self.pop_random_point(), self.grid)

    def pop_random_point(self):
        if len(self.unoccupied_points) == 0:
            return None
        index = random.randrange(len(self.unoccupied_points))
        return self.unoccupied_points.pop(index)

    def save_file(self, filename):
        data = [self.grid, self.turn]
        with open(filename, 'wb') as file:
            pickle.dump(data, file)

            self.screen.clear()
            Logger.log('Saved successfully to: ' + filename)
            Logger.display_logs()

    def load_file(self, filename):
        with open(filename, 'rb') as file:
            game_state = pickle.load(file)
            self.grid = game_state[0]
            self.turn = game_state[1]
            self.player = self.grid.get_human()
            self.grid_map = GridMap(self.grid)
            self.screen = Display(self.grid_map)

            Logger.clear()
            self.screen.clear()

            Logger.log('Loaded successfully from: ' + filename)
            Logger.display_logs()

    def play(self):
        clock = pygame.time.Clock()

        running = True
        reading_new_org = False
        new_org_keys = {K_1: 1, K_2: 2, K_3: 3, K_4: 4, K_5: 5, K_6: 6, K_7: 7, K_8: 8, K_9: 9, K_u: 10, K_i: 11}
        cell = None

        while running:
            # for making action with human
            key = None
            filename_load = ''
            filename_save = ''
            # for clicking buttons and cells
            mouse_pos = pygame.mouse.get_pos()
            for event in pygame.event.get():
                if event.type == KEYDOWN and event.key == K_ESCAPE:
                    running = False
                elif event.type == QUIT:
                    running = False
                elif event.type == KEYUP:
                    key = event.key
                    if reading_new_org and key in new_org_keys.keys():
                        if self.grid.get_organism(Point(*cell)) is not None:
                            Logger.log('Cell already occupied')
                        else:
                            self.create_organism(Species(new_org_keys[key]), Point(*cell), self.grid)
                            Logger.log('Added new organism: ' + Species(new_org_keys[key]).name)
                        self.screen.clear()
                        Logger.display_logs()
                        reading_new_org = False
                elif event.type == MOUSEBUTTONDOWN and event.button == 1:
                    for button in (self.screen.load_button, self.screen.save_button):
                        if button.rect.collidepoint(mouse_pos):
                            button.clicked = True
                        else:
                            button.clicked = False
                elif event.type == MOUSEBUTTONUP and event.button == 1:
                    if self.screen.save_button.clicked:
                        filename_save = self.screen.save_button.get_filename()
                    elif self.screen.load_button.clicked:
                        filename_load = self.screen.load_button.get_filename()
                    cell = self.grid_map.get_clicked_cell(mouse_pos, self.grid)
                    if cell:
                        self.screen.display_choose_surf()
                        reading_new_org = True

            if len(filename_load) > 0:
                self.load_file(filename_load)
            elif len(filename_save) > 0:
                self.save_file(filename_save)
            # Make move for all organisms if the valid key was pressed
            elif key in type(self).VALID_KEYS:
                reading_new_org = False
                self.screen.clear()
                Logger.log('turn: ' + str(self.turn))
                organisms = sorted(self.grid.get_all_organisms(), key=lambda x: (x.initiative, x.age), reverse=True)
                self.player.update(key, self.grid, self.turn)
                for organism in organisms:
                    if organism is not None and self.grid.get_organism(organism.position) == organism:
                        organism.make_action(self.grid)

                Logger.display_logs()
                self.turn += 1

                if self.grid.get_organism(self.player.position) is not self.player:
                    running = False
            elif key == K_r:
                self.screen.clear()
                self.player.activate_ability(self.turn)
                Logger.display_logs()

            self.screen.update(self.grid)

            clock.tick(60)

File: cpp/src/World.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include "World.h"
#include "Wolf.h"
#include "Sheep.h"
#include "Fox.h"
#include "Turtle.h"
#include "Antelope.h"
#include "Grass.h"
#include "SowThistle.h"
#include "Guarana.h"
#include "Belladonna.h"
#include "SosnowskyHogweed.h"
#include <algorithm> // for sort()
#include <unistd.h> // for sleep()

const int END_WAIT_TIME = 5; // in seconds
const char* GAME_STATES_PATH = "../states/";


World::World() : width(GRID_WIDTH), height(GRID_HEIGHT),
grid(GRID_HEIGHT, std::vector<organismType>(GRID_WIDTH)), out(height, width) {
    for (int i = 0; i < height; ++i)
        for (int j = 0; j < width; ++j)
            grid[i][j] = DEFAULT;

    is_end = false;
    turn = 1;
}

void World::startGame() {
    in.getWindow(out.getGameWindow());

    for (int i = 0; i < height; ++i)
        for (int j = 0; j < width; ++j)
            unoccupied_points.emplace_back(i, j);

    // create human
    Point p = _getRandomPoint();
    human_tmp = std::make_unique<Human>(p.y, p.x);
    human = human_tmp.get();
    grid[p.y][p.x] = HUMAN;
    organisms.push_back(std::move(human_tmp));

    _addEachOrganism();
    _addEachOrganism();
    _addEachOrganism();
    _addEachOrganism();
    _addEachOrganism();
    _addEachOrganism();
    // _addOrganism(ANTELOPE);
    // _addOrganism(TURTLE);
    // _addOrganism(TURTLE);
    // _addOrganism(TURTLE);
    // _addOrganism(SHEEP);
    // _addOrganism(GUARANA);

    std::sort(organisms.begin(), organisms.end(), compareOrganisms);
}

void World::makeTurn() {
    action key;
    key = in.readKey();
    if (key != NOTHING) {
        out.clearInfoBox();
        Display::printMessage("turn " + std::to_string(turn));
    }
    if (key == QUIT_GAME)
        is_end = true;
    else if (key == MOVE_DOWN || key == MOVE_UP || key == MOVE_LEFT || key == MOVE_RIGHT) {
        ++turn;
        human->getAction(key, turn);

        // save the size, because the vector will change while iterating over it
        int size = organisms.size();
        for (int i = 0; i < size; ++i) {
            if (organisms[i]) {
                organisms[i]->increaseAge();
                organisms[i]->makeAction(grid, organisms);
            }
        }
        // remove all nullptr from the vector
        organisms.erase(std::remove(organisms.begin(), organisms.end(), nullptr), organisms.end());

        // check if human still exists
        if (!_humanExists()) {
            is_end = true;
            drawWorld();
            Display::clearInfoBox();
            Display::printMessage("You died. The end.");
            Display::printMessage("Number of turns survived: " + std::to_string(turn - 2));
            // getch();
            sleep(END_WAIT_TIME);
        }
        // check if you won
        else if (_humanExists() && organisms.size() == 1) {
            is_end = true;
            drawWorld();
            Display::clearInfoBox();
            Display::printMessage("Congratulations! You won!");
            Display::printMessage("Number of turns survived: " + std::to_string(turn - 1));
            sleep(END_WAIT_TIME);
            // getch();
        }
        else if (_humanExists() && organisms.size() == 0)
            throw std::invalid_argument("Human somehow alive");


        // sorting
        if (organisms.size() > 0)
            std::sort(organisms.begin(), organisms.end(), compareOrganisms);
    }
    else if (key == USE_ABILITY) {
        human->purify(grid, organisms, turn);
    }
    else if (key == SAVE_FILE) {
        std::string filename = in.getFilename();
        if (_serialize(filename)) {
            out.refreshAllWindows();
            Display::printMessage("Save game state to: " + filename);
        }
        else
            Display::printMessage("Couldn't open the file: " + filename);
    }
    else if (key == LOAD_FILE) {
        std::string filename = in.getFilename();
        if (_deserialize(filename)) {
            out.refreshAllWindows();
            Display::printMessage("Loaded game state from: " + filename);
        }
        else
            Display::printMessage("Couldn't open the file: " + filename);
    }
}

void World::drawWorld() {
    out.clearGameWindow();
    for (std::size_t i = 0; i < organisms.size(); ++i)
        organisms[i]->draw(out);
    out.refreshGameWindow();
}

bool World::isEnd() {
    return is_end;
}

bool World::_humanExists() {
    for (int i = 0; i < organisms.size(); ++i) {
        if (organisms[i] && organisms[i]->getType() == HUMAN)
            return true;
    }
    return false;
}

void World::_addEachOrganism() {
    for (int i = 2; i < 1 + PLANTS + ANIMALS; ++i)
        _addOrganism(static_cast<organismType>(i));
}

void World::_addOrganism(organismType type) {
    Point p = _getRandomPoint();
    grid[p.y][p.x] = type;
    std::unique_ptr<Organism> org = _createOrganism(type, p.y, p.x);
    organisms.push_back(std::move(org));
}

Point World::_getRandomPoint() {
    int max = unoccupied_points.size();
    if (max == 0) {
        throw std::logic_error("no more free points on the map");
    }
    Point result;
    int n = rand() % max;

    auto it = std::next(unoccupied_points.begin(), n); // Get an iterator to the n-th element
    result = std::move(*it); // Save the n-th element to a variable
    unoccupied_points.erase(it); // Remove the n-th element from the list

    return result;
}

bool compareOrganisms(const std::unique_ptr<Organism>& a, const std::unique_ptr<Organism>& b) {
    // sort by initiative
    if (a->getInitiative() != b->getInitiative()) {
        return a->getInitiative() > b->getInitiative();
    }
    // sort by age if initiative is the same
    return a->getAge() > b->getAge();
}

bool World::_serialize(std::string filename) const {
    std::ofstream outfile;
    // file.open(filename, std::ios::binary);
    outfile.open(GAME_STATES_PATH + filename);
    if (outfile.is_open()) {
        // save basic properties of the world
        outfile << width << " " << height << " " << is_end << " " << turn << std::endl;

        // save the grid
        for (const auto& row : grid) {
            for (const auto& elem : row) {
                outfile << elem << " ";
            }
            outfile << std::endl;
        }

        // save all the organisms
        outfile << organisms.size() << std::endl;
        for (const auto& organism : organisms) {
            outfile << *organism << std::endl;
        }

        outfile.close();
        return true;
    }
    else {
        return false;
    }
}

bool World::_deserialize(std::string filename) {
    std::ifstream infile(GAME_STATES_PATH + filename);

    if (infile.is_open()) {
        // clear the previous game state
        organisms.clear();
        grid.clear();

        // load basic variables
        infile >> width >> height >> is_end >> turn;

        grid.resize(height, std::vector<organismType>(width));

        // load the grid
        int org_type;
        organismType type;
        for (int i = 0; i < height; ++i) {
            for (int j = 0; j < width; ++j) {
                infile >> org_type;
                type = static_cast<organismType>(org_type);
                grid[i][j] = type;
            }
        }

        // load the organisms
        int num_of_organisms, y, x;
        infile >> num_of_organisms;
        for (int i = 0; i < num_of_organisms; ++i) {
            infile >> org_type >> y >> x;
            type = static_cast<organismType>(org_type);
            std::unique_ptr<Organism> new_org = _createOrganism(type, y, x);
            infile >> *new_org;
            if (type == HUMAN) {
                // human = static_cast<Human*>(new_org.get());
                human = dynamic_cast<Human*>(new_org.get());
                if (human == nullptr)
                    throw std::invalid_argument("Human incorrectly casted");
            }
            organisms.push_back(std::move(new_org));
        }

        // change out width and height
        out.resizeGameWindow(height, width);
        in.getWindow(out.getGameWindow());

        infile.close();
        return true;
    }
    else {
        return false;
    }
}


std::ostream& operator<<(std::ostream& os, const Organism& org) {
    os << org.getType() << " " << org.getY() << " " << org.getX();
    os << " " << org.getAge() << " " << org.getStrength();

    if (org.getType() == HUMAN) {
        const Human& human = static_cast<const Human&>(org);
        os << " " << human;
    }
    else if (!org.isAnimal()) {
        const Plant& plant = static_cast<const Plant&>(org);
        os << " " << plant;
    }

    return os;
}

std::ostream& operator<<(std::ostream& os, const Human& h) {
    os << h._ability_on << " " << h._last_ability;

    return os;
}

std::ostream& operator<<(std::ostream& os, const Plant& p) {
    os << p._attempts_to_spread;

    return os;
}

std::istream& operator>>(std::istream& in, Organism& org) {
    int age, strength;

    in >> age >> strength;
    org.setStrength(strength);
    org.setAge(age);
    if (org.getType() == HUMAN) {
        Human& human = static_cast<Human&>(org);
        in >> human;
    }
    else if (!org.isAnimal()) {
        Plant& plant = static_cast<Plant&>(org);
        in >> plant;
    }
    return in;
}

std::istream& operator>>(std::istream& in, Human& h) {
    in >> h._ability_on >> h._last_ability;

    return in;
}

std::istream& operator>>(std::istream& in, Plant& p) {
    in >> p._attempts_to_spread;

    return in;
}

std::unique_ptr<Organism> World::_createOrganism(organismType type, int y, int x) {
    switch(type) {
        case HUMAN:
            return std::make_unique<Human>(y, x);
        case WOLF:
            return std::make_unique<Wolf>(y, x);
        case SHEEP:
            return std::make_unique<Sheep>(y, x);
        case FOX:
            return std::make_unique<Fox>(y, x);
        case TURTLE:
            return std::make_unique<Turtle>(y, x);
        case ANTELOPE:
            return std::make_unique<Antelope>(y, x);
        case GRASS:
            return std::make_unique<Grass>(y, x);
        case SOW_THISTLE:
            return std::make_unique<SowThistle>(y, x);
        case GUARANA:
            return std::make_unique<Guarana>(y, x);
        case BELLADONNA:
            return std::make_unique<Belladonna>(y, x);
        case SOS_HOGWEED:
            return std::make_unique<SosnowskyHogweed>(y, x);
        case DEFAULT:
            throw std::invalid_argument("Given wrong type to create organism");
    }
}
